runtime/src/network/deterministic.rs

```diff
@@ -0,0 +1,180 @@
+use crate::{deterministic::Auditor, mocks, Error};
+use futures::{channel::mpsc, SinkExt as _, StreamExt as _};
+use std::{
+    collections::HashMap,
+    net::{IpAddr, Ipv4Addr, SocketAddr},
+    ops::Range,
+    sync::{Arc, Mutex},
+};
+
+/// Range of ephemeral ports assigned to dialers.
+const EPHEMERAL_PORT_RANGE: Range<u16> = 32768..61000;
+
+/// Implementation of [crate::Sink] for a deterministic [Network].
+pub struct Sink {
+    auditor: Arc<Auditor>,
+    me: SocketAddr,
+    peer: SocketAddr,
+    sender: mocks::Sink,
+}
+
+impl crate::Sink for Sink {
+    async fn send(&mut self, msg: &[u8]) -> Result<(), Error> {
+        self.auditor.send(self.me, self.peer, msg);
```

Comment:
@patrick-ogrady 

I want to flag that we're not consistent about whether we audit a function call before/after we check for an error.

For example, here in `send` we always `self.auditor.send` even if the we return `Error::SendFailed`. On the other hand, we only `self.auditor.recv` if we didn't error.

Reply:
(This is not new. The old code looks this way, I just didn't notice this til now.)

Reply:
My intention was to always audit "before" (as soon as the function is called). If you notice anywhere we don't do that, please change to that pattern.

Reply:
https://github.com/commonwarexyz/monorepo/pull/878 changes the pattern to record events before and after, as we discussed